File: sdks/python/tests/test_refs.py
"""Tests for refs.py — sign_refs and verify_refs."""

import pytest

from autonoma.refs import sign_refs, verify_refs


class TestRefs:
    def test_round_trip(self):
        payload = {"testRunId": "abc-123", "refs": {"User": [{"id": 1}]}}
        token = sign_refs(payload, "signing-secret")
        decoded = verify_refs(token, "signing-secret")
        assert decoded == payload

    def test_token_has_three_dot_separated_parts(self):
        token = sign_refs({"foo": "bar"}, "secret")
        parts = token.split(".")
        assert len(parts) == 3

    def test_rejects_wrong_secret(self):
        token = sign_refs({"data": 1}, "correct-secret")
        with pytest.raises(ValueError, match="signature mismatch"):
            verify_refs(token, "wrong-secret")

    def test_rejects_tampered_token(self):
        token = sign_refs({"data": 1}, "secret")
        parts = token.split(".")
        # Tamper with payload
        tampered = parts[0] + "." + parts[1] + "X" + "." + parts[2]
        with pytest.raises(ValueError):
            verify_refs(tampered, "secret")

File: sdks/ruby/test/test_refs.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/autonoma"

class TestRefs < Minitest::Test
  def test_produces_3_part_token
    token = Autonoma::Refs.sign_refs(
      { "refs" => { "userId" => "user-1", "email" => "test@test.com" },
        "testRunId" => "test-run-123", "environment" => "standard" },
      "signing-secret"
    )
    parts = token.split(".")
    assert_equal 3, parts.length
    assert_match(/\A[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\z/, token)
  end

  def test_round_trips_arbitrary_json_refs
    payload = {
      "refs" => { "userId" => "user-1", "nested" => { "count" => 3 } },
      "testRunId" => "test-run-123",
      "environment" => "standard"
    }
    secret = "signing-secret"
    token = Autonoma::Refs.sign_refs(payload, secret)
    got = Autonoma::Refs.verify_refs(token, secret)

    assert_equal "test-run-123", got["testRunId"]
    assert_equal "standard", got["environment"]
    assert_equal "user-1", got["refs"]["userId"]
    assert_equal 3, got["refs"]["nested"]["count"]
  end

  def test_round_trips_canonical_cross_language_token
    token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IlJFRlMifQ.eyJyZWZzIjp7IlVzZXIiOlt7ImlkIjoidXNlci0xIiwiZW1haWwiOiJ0ZXN0QHRlc3QuY29tIn1dfSwidGVzdFJ1bklkIjoidGVzdC1ydW4tMTIzIiwiZW52aXJvbm1lbnQiOiJzdGFuZGFyZCJ9.b2340UY6iXALRK2SaBV0BzZLVbxC8J59_csCUEc-gOw"
    payload = Autonoma::Refs.verify_refs(token, "signing-secret")
    expected = {
      "refs" => { "User" => [{ "id" => "user-1", "email" => "test@test.com" }] },
      "testRunId" => "test-run-123",
      "environment" => "standard"
    }
    assert_equal expected, payload
  end

  def test_rejects_wrong_secret
    token = Autonoma::Refs.sign_refs(
      { "refs" => {}, "testRunId" => "r", "environment" => "e" }, "right-secret"
    )
    assert_raises(RuntimeError) { Autonoma::Refs.verify_refs(token, "wrong-secret") }
  end

  def test_rejects_malformed_token
    assert_raises(RuntimeError) { Autonoma::Refs.verify_refs("only-one-part", "signing-secret") }
  end

  def test_rejects_tampered_payload
    token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IlJFRlMifQ.dGFtcGVyZWQ.b2340UY6iXALRK2SaBV0BzZLVbxC8J59_csCUEc-gOw"
    assert_raises(RuntimeError) { Autonoma::Refs.verify_refs(token, "signing-secret") }
  end
end
